pkg/kube/krt/ambient_test.go

```diff
@@ -0,0 +1,184 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt_test
+
+import (
+	"fmt"
+	"net/netip"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+
+	istio "istio.io/api/networking/v1alpha3"
+	istioclient "istio.io/client-go/pkg/apis/networking/v1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/serviceentry"
+	srkube "istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+// This function has a copy of some functionality from the ambient controller so that we
+// can test that some of the krt functionality works the way we expect
+
+func constructService(_ krt.HandlerContext, svc *corev1.Service) *workloadapi.Service {
+	ports := make([]*workloadapi.Port, 0, len(svc.Spec.Ports))
+	for _, p := range svc.Spec.Ports {
+		ports = append(ports, &workloadapi.Port{
+			ServicePort: uint32(p.Port),
+			TargetPort:  uint32(p.TargetPort.IntVal),
+		})
+	}
+
+	addresses, err := slices.MapErr(getVIPs(svc), func(vip string) (*workloadapi.NetworkAddress, error) {
+		ip, err := netip.ParseAddr(vip)
+		if err != nil {
+			return nil, fmt.Errorf("parse %v: %v", vip, err)
+		}
+		return &workloadapi.NetworkAddress{
+			Network: "testnetwork",
+			Address: ip.AsSlice(),
+		}, nil
+	})
+	if err != nil {
+		log.Warnf("fail to parse service %v: %v", config.NamespacedName(svc), err)
+		return nil
+	}
+
+	// TODO this is only checking one controller - we may be missing service vips for instances in another cluster
+	return &workloadapi.Service{
+		Name:      svc.Name,
+		Namespace: svc.Namespace,
+		Hostname:  string(srkube.ServiceHostname(svc.Name, svc.Namespace, "test.local")),
+		Addresses: addresses,
+		Ports:     ports,
+	}
+}
+
+func constructServiceEntries(_ krt.HandlerContext, svc *istioclient.ServiceEntry) []*workloadapi.Service {
```

Comment:
nit: this is a lot of fake collection to test what we need. Can we not use a simpler one like SimplePodCollection?

Reply:
I'll remove once I can confirm tests and everything hold up on the merging implementation